src/common/WrrQueue.h

```diff
@@ -0,0 +1,344 @@
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef WRR_QUEUE_H
+#define WRR_QUEUE_H
+
+#include "common/Formatter.h"
+#include "common/OpQueue.h"
+
+#include <map>
+#include <list>
+
+/**
+ * Weighted Round Robin queue with stri priority queue
+ *
+ * This queue attempts to be fair to all classes of
+ * operations but is also weighted so that higher classes
+ * get more share of the operations. It is not a strict
+ * weighted round robin, but only approximated to keep
+ * overhead of the algorithm low.
+ */
+
+template <typename T, typename K>
+class WrrQueue : public OpQueue <T, K> {
+  int64_t total_priority;
+
+  typedef std::list<std::pair<unsigned, T> > ListPairs;
+  static unsigned filter_list_pairs(
+    ListPairs *l, std::function<bool (T)> f,
+    std::list<T> *out) {
+    unsigned ret = 0;
+    if (out) {
+      for (typename ListPairs::reverse_iterator i = l->rbegin();
+	   i != l->rend();
+	   ++i) {
+	if (f(i->second)) {
+	  out->push_front(i->second);
+	}
+      }
+    }
+    for (typename ListPairs::iterator i = l->begin();
+	 i != l->end();
+      ) {
+      if (f(i->second)) {
+	l->erase(i++);
+	++ret;
+      } else {
+	++i;
+      }
+    }
+    return ret;
+  }
+
+  struct SubQueue {
+  private:
+    typedef std::map<K, ListPairs> Classes;
+    Classes q;
+    typename Classes::iterator cur;
+    unsigned q_size;
+  public:
+    SubQueue(const SubQueue &other)
+      : q(other.q),
+	cur(q.begin()),
+	q_size(0) {}
+    SubQueue()
+      :	cur(q.begin()),
+	q_size(0) {}
+    void enqueue(K cl, unsigned cost, T item,
+		 bool front = CEPH_OP_QUEUE_BACK) {
+      if (front == CEPH_OP_QUEUE_FRONT) {
+	q[cl].push_front(std::make_pair(cost, item));
+      } else {
+	q[cl].push_back(std::make_pair(cost, item));
+      }
+      if (cur == q.end()) {
+	cur = q.begin();
+      }
+      ++q_size;
+    }
+    std::pair<unsigned, T> front() const {
+      assert(!(q.empty()));
+      assert(cur != q.end());
+      assert(!(cur->second.empty()));
+      return cur->second.front();
+    }
+    void pop_front() {
+      assert(!(q.empty()));
+      assert(cur != q.end());
+      for (typename Classes::iterator i = q.begin(); i != q.end(); ++i) {
+      }
+      assert(!(cur->second.empty()));
+      cur->second.pop_front();
+      if (cur->second.empty()) {
+	q.erase(cur++);
+      } else {
+	++cur;
+      }
+      if (cur == q.end()) {
+	cur = q.begin();
+      }
+      --q_size;
+    }
+    unsigned size() const {
+      return q_size;
+    }
+    bool empty() const {
+      return (q_size == 0) ? true : false;
+    }
+    unsigned remove_by_filter(std::function<bool (T)> f, std::list<T> *out) {
```

Comment:
I would add 'override' here and in the rest of places you are overriding base class function